src/repo.rs

```diff
@@ -490,43 +428,175 @@ impl<T: NetworkAdapter + 'static> Repo<T> {
+        &self,
+        observer: Option<Box<dyn Fn(Vec<DocumentId>) + Send>>,
+    ) -> DocCollection {
+        let (reply, receiver) = bounded(1);
+        self.control_sender
+            .send(ControlEvent::CreateCollection { observer, reply })
+            .unwrap();
+        receiver.recv().unwrap()
+    }
+
+    pub async fn connect_stream<RecvErr, SendErr, S>(
+        &self,
+        collection: CollectionId,
+        stream: S,
+    ) -> Result<(), error::RunConnection<SendErr, RecvErr>>
+    where
+        RecvErr: std::fmt::Debug,
+        SendErr: std::error::Error,
+        S: Sink<NetworkMessage, Error = SendErr> + Stream<Item = Result<NetworkMessage, RecvErr>>,
+    {
+        let (mut sink, stream) = stream.split();
+        let to_send = self.outgoing(collection).map(|m| Ok::<_, SendErr>(m));
+        let sending = to_send.forward(&mut sink).fuse();
+
+        let recv_msg = self.collection_sender.clone();
+        let recving = stream
+            .map_err(error::RunConnection::<SendErr, RecvErr>::Recv)
+            .try_for_each({
+                move |m| {
+                    futures::future::ready(
+                        recv_msg
+                            .try_send((collection, CollectionEvent::Network(m.into())))
+                            .map_err(|_| error::RunConnection::<SendErr, RecvErr>::RecvCollection),
+                    )
+                }
+            })
+            .fuse();
+
+        let finished = futures::future::select(sending, recving).await;
+        match finished {
+            futures::future::Either::Left((result, _recving)) => match result {
+                Ok(_) => {
+                    tracing::trace!("we're closing the connection");
+                    Ok(())
+                }
+                Err(e) => {
+                    tracing::trace!(err=?e, "error while sending");
+                    Err(error::RunConnection::Send(e))
+                }
+            },
+            futures::future::Either::Right((result, _sending)) => {
+                if let Err(e) = sink.close().await {
+                    tracing::trace!(err=?e, "error while closing sink");
+                }
+                match result {
+                    Ok(_) => {
+                        tracing::trace!("the remote closed the connection");
+                        Ok(())
+                    }
+                    Err(e) => {
+                        tracing::trace!(err=?e, "error while receiving");
+                        Err(e)
+                    }
+                }
+            }
+        }
+    }
+
+    fn outgoing(&self, collection: CollectionId) -> impl Stream<Item = NetworkMessage> {
+        OutgoingMessages {
+            outboxes: self.outboxes.clone(),
+            id: collection,
+        }
+    }
+
+    pub fn shutdown(&self) {
+        self.control_sender.send(ControlEvent::Shutdown).unwrap();
+    }
+}
+
+struct OutgoingMessages {
+    outboxes: Arc<Mutex<HashMap<CollectionId, Outbox>>>,
```

Comment:
I think this should be `Arc<Mutex<Outbox>>`. See also comment above about the `RepoHandle`.